### DISCUSSION TITLE
Flow: Is Rollup Started From Parent

### DISCUSSION THREAD

**Discussion by rodgersjosh** [jamessimone/apex-rollup#336] [Q&A] [answered]
Hi -- I have been using DLRS for years, and am really looking forward to the improvements this product makes.
I just installed today, and I am attempting to use a Flow with the 'Is Rollup Started From Parent' set.
I am following the guide at : https://github.com/jamessimone/apex-rollup/wiki/Perform-Rollup-On-Records---Base-Invocable-Action

I am having a bit of difficulties -- not getting any errors, but I am not seeing the results of my rollup.

Here is my scenario:
Parent Table: Lead
"Child" Table: tdc_tsw__Scheduled_Sms__c
Lookup field on child: tdc_tsw__Lead__c

I have included a screen shot of my invocable action config.  If anyone has insight into how I am misconfiguring, it would be greatly appreciated.

![RollupsConfig](https://user-images.githubusercontent.com/13787608/174881378-f017f219-e4f1-42aa-a6c8-24346537bf80.jpg)

truly not sure what I am doing incorrectly.  Debugger shows class running, and I don't see any errors.  Any help would be greatly appreciated!

Note: If I do the same (not via a flow, but an apex trigger on parent table), I can get it to work.  However, my goal is to fire via flow
Note: If I configure it on the child object via a flow, I can get it to work.  However, my goal is to fire on the parent object.

I installed the Custom Object Rollup Logger, but I have not successfully got any logs (not: I don't see 'Is Rollup Logging Enabled?' field)

**Comment by jamessimone** [comment]
@rodgersjosh have you created the Org Default record for the Rollup Settings custom setting? That’s what it sounds like the issue is. Also note that the “Is Rollup Logging Enabled” flag is on the Rollup Control CMDT record, of which one comes by default with the package (conveniently also called “Org Default”).

I’ve also noticed you setting the Rollup Context to `UPDATE` - are you only interested in rolling up updates? Otherwise if your record triggered flow is set to run on record create and update, you’ll want to use `UPSERT` as the context there. Please let me know if that all works, I’m more than happy to assist further! I think you’ll find that once you have the initial setup out of the way, Apex Rollup will make it easy for you to graduate off of DLRS.

**Reply by rodgersjosh** [reply]
@jamessimone : thanks so much for a same day response!  Truly appreciated.  I can tell by the quick response and your thorough write up on apex rollups that it is a passion project -- one I truly admire by the way :)

I will take you up on your offer to help :)  My hope is that I can somehow return the favor one day.

I believe I have the org defaults set up correctly.   I have included a screen shot of them from my org so you can confirm:

![ApexRollup_CustomSettings_OrgWideDefaults](https://user-images.githubusercontent.com/13787608/174959913-51f7cd5c-041c-4c01-9d91-4d6bb6acdaf8.jpg)

Note 1: My thinking is that if the org default was not correct, it wouldn't work when I put the RT on the child table (note: things seem to work when I put the RT on the child table; just not working when RT is on the parent table and I use options: 'Calc Item Type When Rollup Started From Parent' with 'Is Rollup Started From Parent parameters'); however, I will obviously leave this for you to confirm.

Note 2: thanks for the heads up on logging.  With your direction I was able to find the setting.  I was just looking in the wrong place.  Thanks for clarification.

Note 3: You are correct -- I should be using UPSERT as the context.  I was playing around in trying to get it to work.  I have switched back to UPSERT.

thanks again!  Your willingness to help so quickly is admirable!

Unfortunately, I am at the same place as described in the original post.  I don't see any errors (even digging through logs on dev console), but I also don't see my rollup field updated.  

Hoping I am missing something obvious with the Org Default record that you will spot :)

**Reply by jamessimone** [reply]
@rodgersjosh there isn’t anything obviously wrong with your setup, at least not that I can see from what you’ve sent over. Is it possible for you to debug the record triggered flow with an example of a record being inserted and either attach the Apex Debug log that gets generated here (alternatively can you email it to me)? When the rollup logging is enabled, the Apex debug log should contain plenty of info as to what’s going on under the hood.

It’s quite curious that everything works correctly from the parent when setup through Apex but not through Flow, as they use the same code path for almost everything. In the meantime, I’ll look through the flow-specific code to see if anything stands out.

**Reply by rodgersjosh** [reply]
@jamessimone : thanks again.  I hope this is okay -- instead of a record being inserted, I captured the logs on a simple lead update.  I am hoping the below video may help demonstrate my issue and org setup.  (note: in video I try via a flow, and observe it not work.  I then switch to an apex trigger to see it work).

(small tip: I would increase the play speed of video to 1.5)
 
https://www.loom.com/share/b49a88d73d3f4c67b773a5638660aaa4

Debug logs are attached.  I hope I included the correct relevant logs.  I am going to look through these again; however, when I did the same yesterday nothing in particular jumped out at me.  Maybe fresh eyes will help
[apex-07L5i0000029p9eEAA.log](https://github.com/jamessimone/apex-rollup/files/8958699/apex-07L5i0000029p9eEAA.log)
[apex-07L5i0000029p9dEAA.log](https://github.com/jamessimone/apex-rollup/files/8958700/apex-07L5i0000029p9dEAA.log)

**Reply by jamessimone** [reply]
@rodgersjosh OK, I'll probably have to make an update (targeted for later today) to get this fully working for you (with the `UPSERT` context) from flow - in the _meantime_, I would imagine that you can get this working with your _current_ setup by simply flipping the Flow Context to `REFRESH`

**Reply by rodgersjosh** [reply]
@jamessimone  : awesome.  really appreciate it.  You were correct -- If I changed flow context to REFRESH it correctly calculated.

I don't want to add more work for you, but also wanted to let you know--not sure if this is a bug or my misunderstanding

1.  I changed the flow context to REFRESH and observed to correctly calculating rollup on save of lead (parent object)
2.  I performed a delete of all child rows
3.  I again triggered a change on the lead -- however, I didn't see my rollup change.  My expectation is that this would be set to 0 or "nulled out"

Do you think my understanding of expected behavior is correct?

Note: here on our implementation: Our child table is "highly transactional"--which means the data in it changes frequently through either batch processing/scheduled jobs, etc.   For this reason, we are attempting to really gain control over when the the rollup is calculated and do so at the parent object level.

I hope all that makes sense.  If I can provide anymore info, I will be more than glad to provide.

I am located in Oregon, but my company is in Colorado.  The next time I am there I owe you a beer for all your help!

**Reply by jamessimone** [reply]
Let me check on what you’ve outlined with all the child rows being deleted and the parent field not being cleared out. That might be a gap in the REFRESH combined with parent fields combo there. I’ll keep you updated on that one. I try to push out updates for issues like this ASAP.

Always happy to help! Let me know if you’re ever coming through Denver.

**Reply by rodgersjosh** [reply]
@jamessimone : very cool.  I really appreciate it.  Just let me know when I should grab your modified package and/or where I should grab it from.  If you want me to serve as a beta tester before making it publicly available, I readily volunteer!
I am really loving the the product.

**Reply by jamessimone** [reply]
@rodgersjosh I will let you know - I don't think I will need any beta testing for these changes, which means that as soon as #337 is merged, the updated package info will show up on the README that you see by default when visiting the "homepage" of this github. I expect to have the updates released by tomorrow, or Friday latest.

**Reply by jamessimone** [reply]
@rodgersjosh just keeping you updated here - [there is currently an issue with packaging](https://trailblazer.salesforce.com/issues_view?Id=a1p4V0000013tex) that I am affected by. My colleagues expect to have this fixed within the next 24 hours. Once I'm able to package again, I'll be able to push out the fixes I've talked about here.

**Reply by rodgersjosh** [reply]
@jamessimone : thanks for keeping me updated.  much appreciated.  Looking forward to the fixes!

**Reply by jamessimone** [reply]
@rodgersjosh the new version is out!

**Comment by rodgersjosh** [comment, marked as answer]
Awesome! Thanks so much for the quick fix and deployment!

I will get this installed and tested asap and let you know.

Josh

**Reply by rodgersjosh** [reply]
@jamessimone : just wanted to give you final confirmation -- I installed the new version and saw both "bugs" corrected.
I changed the context back to upsert and saw it properly calc my rollup from parent.  I then deleted "child" rows, saved parent and saw it clear out the rollup  (side note: cleared field, did not make it 0, which is just fine by us 😁

**Reply by rodgersjosh** [reply]
@jamessimone : thanks again! truly awesome utility.

**Reply by jamessimone** [reply]
No problem! Happy to help. You know where to find me if you need something else. And yes, after some waffling about how to best clear fields (and whether or not numbers should “reset” to 0 or null), I decided to align with how SOQL would handle 0 records for an aggregate with null instead of 0.

**Reply by rodgersjosh** [reply]
@jamessimone : thanks again for the help.  Unfortunately, we are seeing some odd behavior when our flow with the rollup action is activated.  

Note here: I feel crazy even suggesting that this is due to the apex rollup -- I know how obscure/unrelated/crazy this sounds 😊:

1. I am updating an unrelated field on the parent object(lead) 
2. When my "rollup flow" is activated, I regularly (though not always) see unrelated field updates to the parent object (lead) rollback and/or not commit
3.  Because it is an field update not even related to my rollup, it doesn't feel like 'apex rollup' is the culprit.  
4. However, if I deactivate this flow containing the apex rollup, I do not observe this behavior.

My only thread of suspicion is do to the 'Record__Prior' parameter to apex rollup.  This would have my old value that I see persisting despite my update. (side note: I keep asking myself why this parameter is necessary...)

Again--I feel bad even asking, but I am wondering if you have seen something like this before?

The video below demonstrates this as best I could.  Given that their is an async component to this, you will notice that I do not reproduce everytime (timing) -- however, it is somewhat regular to see the unrelated field values from the parent rollback when the RT Flow is activated.  However, I have yet to reproduce when the flow is deactivated.

https://www.loom.com/share/4433c9aaab96484e871f3d0ce76e99c0

(side note: I know it is hard to believe, but I have dug through this orgs other automation components to see if they are the culprit.  However, the only time I am not able to regularly reproduce the issue is when my flow calling the apex rollup is deactivated).

Thanks again.  I really appreciate the help.

**Reply by jamessimone** [reply]
@rodgersjosh Ah, I see the issue. This will be an easy fix - my apologies. I will be pushing out an update shortly.

**Reply by jamessimone** [reply]
@rodgersjosh - I've [created a beta package version](https://test.salesforce.com/packaging/installPackage.apexp?p0=04t6g000008SkX1AAK) which I'd love to verify fixes the issue for you. Would you be able to test this out?

**Comment by rodgersjosh** [comment]
For sure! More than happy to test out

Just let me know the url and I will get it installed and tested

**Reply by jamessimone** [reply]
The URL was included with my last comment - note that beta packages can only be installed to sandboxes/scratch orgs, and not production. Let me know if that's going to be an issue. Otherwise, let me know the results!

**Reply by rodgersjosh** [reply]
@jamessimone : sorry for the delayed reply.  I had replied with a follow up on Friday night (via email), but I don't see it posted here--my apologies.  copying and pasting my reply below:

"fyi, I was able to get this installed on a sandbox and configured. 

Our overall configuration on that sandbox is a bit different, but I configured a flow just like our other environment to utilize the invocable action apex rollups provides.  Rollup worked like a charm, and I didn't see any rollbacks!  I will do some more testing, but I think things look good."

I did some more testing, and things seemed to work perfectly.
Would love to get the prod url with this fix so we could apply to our other environment!  Thanks again

**Reply by jamessimone** [reply]
@rodgersjosh oh dear, apologies for the delay then! I wonder what happened to that original response 🤔. I will get this released today.

**Reply by jamessimone** [reply]
@rodgersjosh this should be all set

**Comment by rodgersjosh** [comment]
hey James -- fyi, I was able to get this installed on a sandbox and
configured.

Our overall configuration on that sandbox is a bit different, but I
configured a flow just like our other environment to utilize the invocable
action apex rollups provides.  Rollup worked like a charm, and I didn't see
any rollbacks!  I will do some more testing, but I think things look good.

**Comment by rodgersjosh** [comment]
Whoops, sorry—replying from my phone so I didn’t realize the link. Thanks
for clarification

We were working out of a dev org so that would be ideal to test from,  but
no issue. I can try and find an equivalent sandbox and let you know. Might
not be able to let you know until tomorrow.

Thanks again for the quick turnaround—greatly appreciated

**Comment by jamessimone** [comment]
Awesome - I’m really happy to hear it!
